File: ecell/libecs/FullID.cpp
//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
//
//       This file is part of the E-Cell System
//
//       Copyright (C) 1996-2016 Keio University
//       Copyright (C) 2008-2016 RIKEN
//       Copyright (C) 2005-2009 The Molecular Sciences Institute
//
//::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
//
//
// E-Cell System is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public
// License as published by the Free Software Foundation; either
// version 2 of the License, or (at your option) any later version.
// 
// E-Cell System is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public
// License along with E-Cell System -- see the file COPYING.
// If not, write to the Free Software Foundation, Inc.,
// 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
// 
//END_HEADER
//
// written by Koichi Takahashi <shafi@e-cell.org>,
// E-Cell Project.
//

#ifdef HAVE_CONFIG_H
#include "ecell_config.h"
#endif /* HAVE_CONFIG_H */

#include <string>

#include "Util.hpp"
#include "Exceptions.hpp"

#include "FullID.hpp"

namespace libecs
{

///////////////////////    SystemPath

void SystemPath::parse( String const& systempathstring )
{
    if( systempathstring.empty() )
    {
        return;
    }

    String aString( systempathstring );
    eraseWhiteSpaces( aString );
    
    String::size_type aCompStart( 0 );
    
    // absolute path ( start with '/' )
    if( aString[ 0 ] == DELIMITER )
    {
        push_back( String( 1, DELIMITER ) );
        ++aCompStart;
    }

    for ( String::size_type aCompEnd; aCompStart < aString.size();
            aCompStart = aCompEnd + 1 )
    {
        aCompEnd = aString.find_first_of( DELIMITER, aCompStart );
        if ( aCompEnd == String::npos )
        {
            aCompEnd = aString.size();
        }

        String aComponent( aString.substr( aCompStart, aCompEnd - aCompStart ) );
        if ( aComponent == ".." )
        {
            if ( theComponents.size() == 1 &&
                    theComponents.front()[ 0 ] == DELIMITER )
            {
                THROW_EXCEPTION( BadSystemPath, 
                                 "Too many levels of retraction with .." );
            }
            else if ( theComponents.empty() || theComponents.back() == ".." )
            {
                theComponents.push_back( aComponent );
            }
            else 
            {
                theComponents.pop_back();
            }
        }
        else if ( !aComponent.empty() && aComponent != "." )
        {
            theComponents.push_back( aComponent );
        }
    }

    if ( !aString.empty() && theComponents.empty() )
    {
        theComponents.push_back( "." );
    }
}

String SystemPath::asString() const
{
    StringVector::const_iterator i = theComponents.begin();
    String aString;

    if( isAbsolute() )
    {
        if( theComponents.size() == 1 )
        {
            return "/";
        }
        else
        {
            ; // do nothing
        }
    }
    else
    {
        // isAbsolute() == false implies that this can be empty
        if( theComponents.empty() )
        {
            return aString;
        }
        else
        {
            aString = *i;
        }
    }

    if( i == theComponents.end() )
    {
        return aString;
    }

    ++i;

    while( i != theComponents.end() )
    {
        aString += '/';
        aString += *i;
        ++i;
    }

    return aString;
}

void SystemPath::canonicalize()
{
    StringVector aNewPathComponents;

    for ( StringVector::const_iterator i( theComponents.begin() );
           i != theComponents.end(); ++i )
    {
        if ( *i == "." )
        {
            continue;
        }
        else if ( *i == ".." )
        {
            if ( aNewPathComponents.empty() )
            {
                break;
            }
            aNewPathComponents.pop_back();
        }
        else
        {
            aNewPathComponents.push_back( *i );
        }
    }

    theComponents.swap( aNewPathComponents );
}

SystemPath SystemPath::toRelative( SystemPath const& aBaseSystemPath ) const
{
    // 1. "" (empty) means Model itself, which is invalid for this method.
    // 2. Not absolute is invalid (not absolute implies not empty).
    if( ! isAbsolute() || isModel() )
    {
        return *this;
    }

    if( ! aBaseSystemPath.isAbsolute() || aBaseSystemPath.isModel() )
    {
        THROW_EXCEPTION( BadSystemPath, 
                         "[" + aBaseSystemPath.asString() +
                         "] is not an absolute SystemPath" );
    }

    SystemPath aThisPathCopy;
    SystemPath const* thisPath;

    if ( !isCanonicalized() )
    {
        aThisPathCopy = *this;
        aThisPathCopy.canonicalize();
        thisPath = &aThisPathCopy;
    }
    else
    {
        thisPath = this;
    }

    SystemPath aBaseSystemPathCopy;
    SystemPath const* aCanonicalizedBaseSystemPath;
    if ( !aBaseSystemPath.isCanonicalized() )
    {
        aCanonicalizedBaseSystemPath = &aBaseSystemPath;
    }
    else
    {
        aBaseSystemPathCopy = aBaseSystemPath;
        aBaseSystemPathCopy.canonicalize();
        aCanonicalizedBaseSystemPath = &aBaseSystemPathCopy;
    }

    SystemPath aRetval;
    StringVector::const_iterator j( thisPath->theComponents.begin() ),
                               je( thisPath->theComponents.end() );
    StringVector::const_iterator
        i( aCanonicalizedBaseSystemPath->theComponents.begin() ),
        ie( aCanonicalizedBaseSystemPath->theComponents.end() );

    while ( i != ie && j != je )
    {
        String const& aComp( *i );
        if ( aComp != *j )
        {
            break;
        }
        ++i, ++j;
    }
    if ( i != ie )
    {
        while ( i != ie )
        {
            aRetval.theComponents.push_back( ".." );
            ++i;
        }
    }
    std::copy( j, je, std::back_inserter( aRetval.theComponents ) );

    if ( aRetval.theComponents.empty() )
    {
        aRetval.theComponents.push_back( "." );
    }

    return aRetval; 
}


///////////////// FullID

void FullID::parse( String const& fullidstring )
{
    // empty FullID string is invalid
    if( fullidstring == "" )
    {
        THROW_EXCEPTION( BadID, "empty FullID string" );
    }

    String aString( fullidstring );
    eraseWhiteSpaces( aString );

    // ignore leading white spaces
    String::size_type aFieldStart( 0 );
    String::size_type aFieldEnd( aString.find_first_of( DELIMITER, aFieldStart ) );
    if( aFieldEnd == String::npos )
    {
        THROW_EXCEPTION( BadID, 
                         "no ':' in the FullID string [" + aString + "]" );
    }

    String aTypeString( aString.substr( aFieldStart, aFieldEnd - aFieldStart ) );
    theEntityType = EntityType( aTypeString );
    
    aFieldStart = aFieldEnd + 1;
    aFieldEnd = aString.find_first_of( DELIMITER, aFieldStart );
    if( aFieldEnd == String::npos )
    {
        THROW_EXCEPTION( BadID, "only one ':' in the FullID string [" 
                                + aString + "]" );
    }

    
    theSystemPath = SystemPath( aString.substr( aFieldStart, 
                                aFieldEnd - aFieldStart ) );
    
    aFieldStart = aFieldEnd + 1;

    // drop trailing string after extra ':'(if this is    FullPN),
    // or go to the end
    aFieldEnd = aString.find_first_of( DELIMITER, aFieldStart );

    theID = aString.substr( aFieldStart, aFieldEnd - aFieldStart );
}

String FullID::asString() const
{
    if ( theID.empty() )
    {
        return String( "(invalid)" );
    }

    return theEntityType.asString() + FullID::DELIMITER 
        + theSystemPath.asString() + FullID::DELIMITER + theID;
}

bool FullID::isValid() const
{
    bool aFlag( theSystemPath.isValid() );
    aFlag &= ! theID.empty();

    return aFlag;
}


///////////////// FullPN


FullPN::FullPN( String const& fullpropertynamestring )
    : theFullID( fullpropertynamestring )
{

    String::size_type aPosition( 0 );

    for( int i( 0 ) ; i < 3 ; ++i )
    {
        aPosition = fullpropertynamestring.
            find_first_of( FullID::DELIMITER, aPosition );
        if( aPosition == String::npos ) 
        {
            THROW_EXCEPTION( BadID, "not enough fields in FullPN string [" +
                                    fullpropertynamestring + "]" );
        }
        ++aPosition;
    }

    thePropertyName = fullpropertynamestring.substr( aPosition, String::npos );
    eraseWhiteSpaces( thePropertyName );
}

String FullPN::asString() const
{
    return theFullID.asString() + FullID::DELIMITER + thePropertyName;
}

bool FullPN::isValid() const
{
    return theFullID.isValid() & ! thePropertyName.empty();
}

} // namespace libecs

File: ecell/pyecell/ecell/FullID.py
#!/usr/bin/env python
#::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
#
#       This file is part of the E-Cell System
#
#       Copyright (C) 1996-2010 Keio University
#       Copyright (C) 2005-2009 The Molecular Sciences Institute
#
#::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
#
#
# E-Cell System is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public
# License as published by the Free Software Foundation; either
# version 2 of the License, or (at your option) any later version.
# 
# E-Cell System is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
# See the GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public
# License along with E-Cell System -- see the file COPYING.
# If not, write to the Free Software Foundation, Inc.,
# 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
# 
#END_HEADER

from ecs_constants import *

import types

class FullID( tuple ):
    def __init__( self , aSequence ):
        tuple.__init__( self , aSequence )
        self.validateLength()

    def validateLength( self ):
        # over-written in FullPN
        if len( self ) != 3:
            raise ValueError( 'FullID has 3 fields. ( %d given )'
                              % len( aSequence ) )

    def getEntityTYpeString( self ):
        return ENTITYTYPE_STRING_LIST[ int( self[ TYPE ] ) ]

    def getSystemPathString( self ):
        return self[ SYSTEMPATH ].getString()
    
    def getFullID( self ):
        return createFullID( self[ : 3 ] )

    def getFullPN( self , aPropertyName='' ):
        if len( self ) == 4:
            if aPropertyName == '':
                aPropertyName = self[ PROPERTY ]
        return createFullPN( tuple( self[ : 3 ] ) + ( aPropertyName , ) )

    def getString( self ):
        aTypeString = self.getEntityTYpeString()
        aSystemPathString = self.getSystemPathString()
        return aTypeString + ':' + \
               aSystemPathString + ':' + \
               ':'.join( self[ 2 : ] )

    def convertToSystemPath( self ):
        if self[ TYPE ] == SYSTEM:
            return createSystemPath( self[ SYSTEMPATH ] + ( self[ ID ] , ) )
        else:
            raise TypeError( 'can not convert FullID of %s type entity'
                             ' to SystemPath' % self.getEntityTYpeString() )

    def getSuperSystemFullID( self ):
        return self[ SYSTEMPATH ].getFullID()

class FullPN( FullID ):
    def __init__( self , aSequence ):
        FullID.__init__( self , aSequence )

    def validateLength( self ):
        if len( self ) != 4:
            raise ValueError( 'FullPN has 4 fields. ( %d given ).'
                              % len( aSequence ) )

class SystemPath( tuple ):
    def __init__( self , aSequence ):
        tuple.__init__( self , aSequence )

    def getFullID( self ):
        return createFullID( ( SYSTEM , self[ : -1 ] , self[ -1 ] ) )

    def getFullPN( self , aPropertyName = '' ):
        return self.getFullID().getFullPN( aPropertyName )

    def getString( self ):
        return '/' + '/'.join( self[ 1 : ] )

    def isAbsolute( self ):
        if self[0] == '/':
            return 1
        else :
            return 0

    def joinSystemPath( self , aRelativeSystemPath ):
        if aRelativeSystemPath.isAbsolute():
            raise ValueError( 'can not join absolute system path' )

        aSystemPath = createSystemPath( self + aRelativeSystemPath )
        aList = aSystemPath.solveParentReference()
        
        return createSystemPath( aList )

    def solveParentReference( self , aList = None ):
        if not aList:
            aList = list( self )

        if aList[0] == '..':
            aList = self.solveParentReference( aList[ 1 : ] )
            return ['..'] + aList

        try:
            aPosition = aList.index( '..' )
        except ValueError:
            return aList

        if aList[ aPosition - 1 ] == '/':
            raise ValueError( 'root system has no parent system' )
        
        aNewList = aList[ : aPosition - 1 ] + aList[ aPosition + 1 : ]
        aList = self.solveParentReference( aNewList )
        return aList
        
def createFullID( aValue ):
    aList = convertToFullIDList( aValue )
    if not aList:
        raise TypeError( 'can not create FullID from %s type object'
                         % type( aValue ) )
    else:
        aList[ SYSTEMPATH ] = createSystemPath( aList[ SYSTEMPATH ] )
        return FullID( aList )

def createFullPN( aValue ):
    aList = convertToFullIDList( aValue )
    if not aList:
        raise TypeError( 'can not create FullPN from %s type object'
                         % type( aValue ) )
    else:
        aList[ SYSTEMPATH ] = createSystemPath( aList[ SYSTEMPATH ] )
        return FullPN( aList )

def createSystemPath( aValue ):
    # empty system path
    if aValue == '':
        return SystemPath( () )

    aList = convertToSystemPathList( aValue )
    if not aList:
        raise TypeError( 'can not create SystemPath from %s type object'
                         % type( aValue ) )

    return SystemPath( aList )

def convertToList( aValue ):
    if isinstance( aValue , list ):
        return aValue
    elif isinstance( aValue , tuple ):
        return list( aValue )
    else:
        return None

def convertToFullIDList( aValue ):
    aList = convertToList( aValue )
    if aList:
        return aList

    else:
        if isinstance( aValue , str ):
            aList = aValue.split( ':' )
            try:
                aList[ 0 ] = ENTITYTYPE_DICT[ aList[ 0 ] ]
            except IndexError:
                raise ValueError( 'Invalid EntityTYpe string (%s).'
                                  % aList[ 0 ] )
            return aList

        else:
            return None

def convertToSystemPathList( aValue ):
    aList = convertToList( aValue )
    if aList:
        return aList

    else:
        if isinstance( aValue , str ):
            aList = aValue.split( '/' )
            if aValue == '':
                return []

            if aList[ 0 ] == '':
                del( aList[ 0 ])

            if aValue.lstrip()[ 0 ] == '/':
                aList[ 0 : 0 ] = ['/']

            return aList

        else:
            return None

if __name__ == '__main__':

    aSystemPath = createSystemPath( ( '/' , 'CELL' , 'CYTOPLASM' ) )
    print aSystemPath

    print

    aTuple = ( SYSTEM , ( '/' , 'CELL' , 'CYTOPLASM' ) , 'CHROMOSOME' )
    aFullID = createFullID( aTuple )
    print aFullID

    aFullID = createFullID( 'System:/CELL/CYTOPLASM:CHROMOSOME' )
    print aFullID

    aFullID = createFullID( ( SYSTEM , aSystemPath , 'CHROMOSOME' ) )
    print aFullID

    print

    aTuple = ( VARIABLE , ( '/' , 'CELL' , 'CYTOPLASM' ) , 'ATP' , 'CONC' )
    aFullPN = createFullPN( aTuple )
    print aFullPN

    aFullPN = createFullPN( 'Variable:/CELL/CYTOPLASM:ATP:CONC' )
    print aFullPN

    print

    print aSystemPath
    print aSystemPath.getFullID()
    print aSystemPath.getFullPN( 'VOLUME' )
    print aSystemPath.getString()

    print

    print aFullID
    print aFullID.getFullPN( 'VOLUME' )
    print aFullID.getFullID()
    print aFullID.getString()
    print aFullID.getSuperSystemFullID()

    try:
        print aFullID.convertToSystemPath()
    except TypeError:
        print 'type error!'


    aFullID = createFullID( 'Variable:/CELL/CYTOPLASM:ATP' )
    try:
        print aFullID.convertToSystemPath()
    except TypeError:
        print 'type error!'

    print

    print aFullPN
    print aFullPN.getFullPN( 'VOLUME' )
    print aFullPN.getFullID()
    print aFullPN.getString()

    print aFullPN.getSuperSystemFullID()

    try:
        print aFullPN.convertToSystemPath()
    except TypeError:
        print 'type error!'

    aFullPN = createFullPN( 'System:/CELL:CYTOPLASM:VOLUME' )

    try:
        print aFullPN.convertToSystemPath()
    except TypeError:
        print 'type error!'

    print 

    print aSystemPath
    print aSystemPath.isAbsolute()
    aRelativeSystemPath = createSystemPath('CHROMOSOME/GENE')
    print aRelativeSystemPath
    print aRelativeSystemPath.isAbsolute()
    print aSystemPath.joinSystemPath( aRelativeSystemPath )

    aRelativeSystemPath = createSystemPath('CHROMOSOME/../GENE')
    print aRelativeSystemPath
    print aSystemPath.joinSystemPath( aRelativeSystemPath )
    
    aRelativeSystemPath = createSystemPath('../../ENVIRONMENT')
    print aRelativeSystemPath
    print aSystemPath.joinSystemPath( aRelativeSystemPath )

    print 
    
    aSystemPath = createSystemPath( 'A/B/C' )
    aRelativeSystemPath = createSystemPath( '../E/F/../../G/H/I/J/..' )
    print aSystemPath
    print aRelativeSystemPath
    print aSystemPath.joinSystemPath( aRelativeSystemPath )
    
    print
    
    aSystemPath = createSystemPath( 'A/B/C' )
    aRelativeSystemPath = createSystemPath( '../../../E/F/../../../../G/H/I/J/..' )
    print aSystemPath
    print aRelativeSystemPath
    print aSystemPath.joinSystemPath( aRelativeSystemPath )
    
    print
    
    aSystemPath = createSystemPath( '' )
    print aSystemPath

    print
    
    print 'end'

    
